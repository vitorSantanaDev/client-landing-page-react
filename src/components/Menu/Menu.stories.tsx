import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import Menu from '.'
import linksMock from '../NavLinks/mock'

export default {
  title: 'Menu',
  component: Menu,
  args: {
    links: linksMock,
    logoData: {
      text: 'Logo',
      link: '#target',
      srcImage: ''
    }
  }
} as ComponentMeta<typeof Menu>

export const Basic: ComponentStory<typeof Menu> = (args) => <Menu {...args} />

Basic.args = {}

export const Default: ComponentStory<typeof Menu> = (args) => <Menu {...args} />
