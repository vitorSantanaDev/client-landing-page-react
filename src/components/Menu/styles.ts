import styled, { css } from 'styled-components'
import { Wrapper as SectionContainer } from '../SectionContainer/styles'
import { Wrapper as Heading } from '../Heading/styles'

interface IWrapper {
  visible: boolean
}

const menuVisible = () => css`
  opacity: 1;
  visibility: visible;
`

export const Wrapper = styled.div<IWrapper>`
  ${({ theme, visible }) => css`
    top: 0;
    left: 0;
    right: 0;
    position: fixed;
    z-index: 5;
    transition: all 300ms ease-in-out;
    border-bottom: ${theme.colors.mediumGray};
    background-color: ${theme.colors.white};

    > ${SectionContainer} {
      padding-top: 0;
      padding-bottom: 0;
    }

    & ${Heading} {
      margin-top: 0;
      margin-bottom: 0;
    }

    @media ${theme.media.lteMedium} {
      opacity: 0;
      height: 100vh;
      visibility: hidden;

      ${visible && menuVisible()}

      > ${SectionContainer} {
        margin: 0;
        height: 100vh;
        overflow-y: auto;

        display: grid;
        align-items: center;
        grid-template-rows: 1fr;
        grid-template-columns: 1fr;
      }

      & ${Heading} {
        padding-bottom: ${theme.spacings.large};

        display: flex;
        justify-content: center;
      }
    }
  `}
`

export const MenuContainer = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: space-between;

    @media ${theme.media.lteMedium} {
      display: block;
      text-align: center;
      padding: ${theme.spacings.xxlarge} 0;
    }
  `}
`

export const Button = styled.button`
  ${({ theme }) => css`
    z-index: 6;
    top: 2rem;
    right: 2rem;
    width: 4rem;
    height: 4rem;
    position: fixed;
    cursor: pointer;
    display: none;

    border: none;
    color: ${theme.colors.white};
    background-color: ${theme.colors.primaryColor};

    > svg {
      width: 2.5rem;
      height: 2.5rem;
    }

    @media ${theme.media.lteMedium} {
      display: flex;
      align-items: center;
      justify-content: center;
    }
  `}
`
