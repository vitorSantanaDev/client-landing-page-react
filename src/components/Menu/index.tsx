import React, { useEffect, useState } from 'react'

import LogoLink from '../LogoLink'
import NavLinks from '../NavLinks'
import SectionContainer from '../SectionContainer'
import { Menu as MenuIcon } from '@styled-icons/material-outlined/Menu'
import { Close as CloseIcon } from '@styled-icons/material-outlined/Close'
import * as S from './styles'

interface IMenuProps {
  links?: {
    children: React.ReactNode
    link: string
    newTab: boolean
  }[]
  logoData: {
    text: string
    srcImage?: string
    link: string
  }
}

const Menu: React.FC<IMenuProps> = ({ links = [], logoData }) => {
  const [menuVisible, setMenuVisible] = useState(false)
  const handleWindowResize = () => {
    setMenuVisible(false)
  }

  useEffect(() => {
    window.addEventListener('resize', handleWindowResize)
    return () => {
      window.removeEventListener('resize', handleWindowResize)
    }
  }, [menuVisible])

  return (
    <>
      <S.Button
        aria-label="Open/Close menu"
        onClick={() => setMenuVisible((s) => !s)}
      >
        {!menuVisible ? (
          <MenuIcon aria-label="Open menu" />
        ) : (
          <CloseIcon aria-label="Close menu" />
        )}
      </S.Button>
      <S.Wrapper onClick={() => setMenuVisible(false)} visible={menuVisible}>
        <SectionContainer>
          <S.MenuContainer>
            <LogoLink {...logoData} />
            <NavLinks links={links} />
          </S.MenuContainer>
        </SectionContainer>
      </S.Wrapper>
    </>
  )
}

export default Menu
