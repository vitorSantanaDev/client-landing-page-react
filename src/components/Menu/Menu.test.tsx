import React from 'react'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderTheme } from '../../styles/theme/render-theme'

import Menu from '.'
import linksMock from '../NavLinks/mock'
import { theme } from '../../styles/theme/theme'

const logoData = {
  text: 'Logo',
  link: '#target',
  srcImage: 'images/logo.svg'
}
describe('<Menu />', () => {
  it('should render Logo and Main menu nav', () => {
    const { container } = renderTheme(
      <Menu logoData={logoData} links={linksMock} />
    )
    const heading = screen.getByRole('heading', { name: 'Logo' })
    const nav = screen.getByRole('navigation', { name: 'Main menu' })

    expect(heading).toBeInTheDocument()
    expect(nav).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render menu mobile and button for open and close the menu', () => {
    renderTheme(<Menu logoData={logoData} links={linksMock} />)
    const button = screen.getByLabelText('Open/Close menu')
    const menuContainer = button.nextSibling

    expect(button).toHaveStyleRule('display', 'none')
    expect(button).toHaveStyleRule('display', 'flex', {
      media: theme.media.lteMedium
    })

    expect(menuContainer).toHaveStyleRule('opacity', '0', {
      media: theme.media.lteMedium
    })
    expect(screen.getByLabelText('Open menu')).toBeInTheDocument()

    userEvent.click(button)
    expect(menuContainer).toHaveStyleRule('opacity', '1', {
      media: theme.media.lteMedium
    })
    expect(screen.getByLabelText('Close menu')).toBeInTheDocument()

    userEvent.click(button)
    expect(menuContainer).toHaveStyleRule('opacity', '0', {
      media: theme.media.lteMedium
    })
    expect(screen.getByLabelText('Open menu')).toBeInTheDocument()
  })

  it('should not render links', () => {
    const { container } = renderTheme(<Menu logoData={logoData} links={[]} />)
    const nav = screen.queryByRole('navigation', {
      name: 'Main menu'
    })?.firstChild

    expect(nav).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
