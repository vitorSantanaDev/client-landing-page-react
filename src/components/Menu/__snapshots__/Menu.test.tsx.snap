// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should not render links 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  z-index: 6;
  top: 2rem;
  right: 2rem;
  width: 4rem;
  height: 4rem;
  position: fixed;
  cursor: pointer;
  display: none;
  border: none;
  color: #FFFFFF;
  background-color: #0A1128;
}

.c0 > svg {
  width: 2.5rem;
  height: 2.5rem;
}

@media (max-width:768px) {

}

@media (max-width:768px) {

}

@media (max-width:768px) {

}

@media (max-width:768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

<button
  aria-label="Open/Close menu"
  class="c0"
>
  <svg
    aria-hidden="true"
    aria-label="Open menu"
    class="c1"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</button>
`;

exports[`<Menu /> should not render menu when window resize 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  z-index: 6;
  top: 2rem;
  right: 2rem;
  width: 4rem;
  height: 4rem;
  position: fixed;
  cursor: pointer;
  display: none;
  border: none;
  color: #FFFFFF;
  background-color: #0A1128;
}

.c0 > svg {
  width: 2.5rem;
  height: 2.5rem;
}

@media (max-width:768px) {

}

@media (max-width:768px) {

}

@media (max-width:768px) {

}

@media (max-width:768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

<button
  aria-label="Open/Close menu"
  class="c0"
>
  <svg
    aria-hidden="true"
    aria-label="Open menu"
    class="c1"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</button>
`;

exports[`<Menu /> should not render nav when menu container is clicked 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  z-index: 6;
  top: 2rem;
  right: 2rem;
  width: 4rem;
  height: 4rem;
  position: fixed;
  cursor: pointer;
  display: none;
  border: none;
  color: #FFFFFF;
  background-color: #0A1128;
}

.c0 > svg {
  width: 2.5rem;
  height: 2.5rem;
}

@media (max-width:768px) {

}

@media (max-width:768px) {

}

@media (max-width:768px) {

}

@media (max-width:768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

<button
  aria-label="Open/Close menu"
  class="c0"
>
  <svg
    aria-hidden="true"
    aria-label="Open menu"
    class="c1"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</button>
`;

exports[`<Menu /> should render Logo and Main menu nav 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  z-index: 6;
  top: 2rem;
  right: 2rem;
  width: 4rem;
  height: 4rem;
  position: fixed;
  cursor: pointer;
  display: none;
  border: none;
  color: #FFFFFF;
  background-color: #0A1128;
}

.c0 > svg {
  width: 2.5rem;
  height: 2.5rem;
}

@media (max-width:768px) {

}

@media (max-width:768px) {

}

@media (max-width:768px) {

}

@media (max-width:768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

<button
  aria-label="Open/Close menu"
  class="c0"
>
  <svg
    aria-hidden="true"
    aria-label="Open menu"
    class="c1"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</button>
`;
